indexing-service/src/main/java/org/apache/druid/indexing/common/task/KillUnusedSegmentsTask.java

```diff
@@ -341,6 +342,15 @@ private NavigableMap<DateTime, List<TaskLock>> getNonRevokedTaskLockMap(TaskActi
     return taskLockMap;
   }
 
+  @Override
+  public boolean isReady(TaskActionClient taskActionClient) throws Exception
+  {
+    if (markAsUnused) {
+      return super.isReady(taskActionClient);
+    }
+    return super.isReady(taskActionClient, getContextValue(Tasks.TASK_LOCK_TYPE, TaskLockType.SHARED));
```

Comment:
@AmatyaAvadhanula @IgorBerman  what would it mean to use shared locks when multiple kill tasks for a datasource are submitted with the same interval or different overlapping intervals? Would the kill tasks result in the same state as if they were run sequentially guaranteeing idempotence in steady state, or could concurrent kill tasks step on each other and cause problems with metadata and deep storage cleanup?

Reply:
@abhishekrb19 Having concurrent kill tasks definitely doesn't seem ideal from a resource perspective.
When using a limit on the number of segments killed, I do not think we can guarantee idempotence either.
We should definitely use EXCLUSIVE as the default and REPLACE as the preferred concurrent lock.

Reply:
>Having concurrent kill tasks definitely doesn't seem ideal from a resource perspective.

Hmm. I was thinking about the plausible scenario where a user submits a micro-batch of kill tasks rather than a large one if a datasource has too many unused segments. That way, the user doesn't have to wait for a long-running task to complete. Right now, if a user did that with two concurrent kill tasks with overlapping intervals, the task that acquires the exclusive lock would do the work and the task that follows it would essentially be a no-op when it gets the chance to run. I'm wondering if that will still be the case?

Re idempotence: in steady state when there are no more newly added unused segments to an interval, I'd think that the kill task should guarantee idempotence even when `limit` and `batchSize` are supplied. Thoughts?

Reply:
If we were to submit two kill tasks with a limit of say 1000 each, isn't it possible that they might try to kill batches that have a few common unused segments, and that way the total number of segments killed is less than 2000?

Reply:
Yeah, I see what you mean. With the follow up discussion to use exclusive lock by default and replace lock in the concurrent case, this won't be a concern anymore, right?